### ISSUE TITLE
Possible regression in AMQP with 1.19.0 with Task handling

### ISSUE THREAD

**Issue by sebader** [Azure/azure-iot-sdk-csharp#807]
- **OS, version, SKU and CPU architecture used:** <VERSION>  Windows 10
- **Application's .NET Target Framework :** <NETVERSION>  NETCORE 2.1
- **Device:** <DEVICETYPE> amd64 
- **SDK version used:** 1.19.0 <VERSION> 

## Description of the issue:

Deriving from this StackOverflow question: https://stackoverflow.com/questions/54759059/azure-iot-hub-cant-call-direct-method-using-amqp-protocol
Without repeating the whole question from there, I believe there is a regression with the SDK 1.19.0. The same code works fine up to 1.18.1. The code also works fine with 1.19.0 with MQTT as transport protocol.

When you change the `Console.ReadKey()` to for example `await Task.Delay(-1)` it works fine. Seems to be something changed in the task handling.

## Code sample exhibiting the issue:

```
using Microsoft.Azure.Devices.Client;
using Microsoft.Azure.Devices.Shared;
using Newtonsoft.Json;
using System;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace VirtualIoTDevice
{
    internal class Program
    {
        private const string DeviceConnectionString = "device-connection-string";
        private const string DEVICE_ID = "device01";
        private static DeviceClient _device;

        private static async Task Main(string[] args)
        {
            Console.WriteLine("Initializing virtual IoT device..");
            using (_device = DeviceClient.CreateFromConnectionString(DeviceConnectionString, DEVICE_ID))
            {
                await _device.OpenAsync();
                await _device.SetMethodHandlerAsync("showMessage", ShowMessage, null);

                Console.ReadKey();
            }
        }

        private static Task<MethodResponse> ShowMessage(MethodRequest methodRequest, object userContext)
        {
            Console.WriteLine("***Direct message received***");
            Console.WriteLine(methodRequest.DataAsJson);

            var responsePayload = Encoding.ASCII.GetBytes(JsonConvert.SerializeObject(new { response = "Message shown!" }));
            return Task.FromResult(new MethodResponse(responsePayload, 200));
        }
    }
}
```

## Console log of the issue:

There is no apparent error/exception. The connection will just never open.

**Comment by davilu** [comment]
@sebader Thanks for reporting the issue. We'll working to get a fix. Before that, as a work around, you could move `Console.ReadKey()` in task like `await Task.Run(() => Console.ReadKey()).ConfigureAwait(false)`

**Comment by jenoola** [comment]
Got a repro, then tried to change LangVersion in csproj: set to an 8.0 (turned out to be unsupported), to 7.0, then to 7.1, 7.2, 7.3, and back to latest. After playing with it, no repro anymore. This looks like a weird compiler bug to me. @CIPop could you comment on this?

**Comment by jenoola** [comment]
Repro with 1.19.0 and 1.20.0 if LangVersion >= 7.1
No repro with 1.19.0 and 1.20.0 if LangVersion is 7.0
No repro with 1.18.1 and any LangVersion
No repro with any SDK and LangVersion if  _device.OpenAsync and _device.SetMethodHandlerAsync are called with Wait() or like this:
```
Task[] tasks = new Task[2] { _device.OpenAsync(), _device.SetMethodHandlerAsync("showMessage", ShowMessage, null) };
Task.WaitAll(tasks);
```
Investigating

**Comment by jenoola** [comment]
Repro with 1.20.1. The error is reported in service :

`{"message":"{\"Message\":\"{\\\"errorCode\\\":504101,\\\"trackingId\\\":\\\"bd60470af6b64cd6b332370351a639d2-G:8-TimeStamp:05/15/2019 18:01:23-G:9-TimeStamp:05/15/2019 18:01:23\\\",\\\"message\\\":\\\"Timed out waiting for the response from device.\\\",\\\"info\\\":{},\\\"timestampUtc\\\":\\\"2019-05-15T18:01:23.4145751Z\\\"}\",\"ExceptionMessage\":\"\"}"}`

**Comment by jenoola** [comment]
@sebader the behaviour you see now is expected. Console.ReadKey() is a blocking I/O call in an async function Main, which means Main cannot return control to the message pump until the blocking call is finished. Since Main keeps the control, no other async functions can proceed. If you make the following change
```
using (_device = DeviceClient.CreateFromConnectionString(DeviceConnectionString, DEVICE_ID))
            {
                await _device.OpenAsync();
                await _device.SetMethodHandlerAsync("showMessage", ShowMessage, null);

                Console.ReadKey(); // 1
                Console.ReadKey(); // 2
            }
```
and then run app until first Console.ReadKey(), invoke DirectMethod from service and press any key in console. At this point ShowMessage will get the control and you will see its output. The second Console.ReadKey() is just to prevent the app from exiting.

Long story short, you can't use blocking calls in async functions. You can use the workaround @davilu suggested or use synchronous Main.

**Comment by az-iot-builder-01** [comment]
@sebader, @davilu, @jenoola, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by sebader** [comment]
@jenoola while I understand what you are saying, I see two issues with that:

1. It does work differently with the same code when MQTT is selected
2. More importantly: This poses a breaking change in a minor SDK update.

**Comment by prmathur-microsoft** [comment]
We will look into this behavior change and get back to you with recomendations. Thanks for highlighting it.

**Comment by CIPop** [comment]
@sebader creating a deadlock within the app code or mixing between app and SDK code is very likely. This is a known fact about async/await TPL operations.
See: https://devblogs.microsoft.com/pfxteam/await-and-ui-and-deadlocks-oh-my/

Relying on us creating a different Task for the method execution was simply covering up the bug (which can still occur within your application's code).
The extra task was removed for perf reasons. Feel free to use @davilu 's recommended workaround above (which is what 1.18 was doing internally).

Closed: by design.
